### PR TITLE
Search actions

### PR DESCRIPTION
**Pull request by VDVsx**

### CODE REVIEW

**Comment by VDVsx** [comment]
@matthewvogt - Plese have a look.

File: src/emailmessagelistmodel.cpp [hunk 1]
```diff
@@ -335,9 +338,14 @@ void EmailMessageListModel::setSearch(const QString search)
...
         QMailMessageKey fromKey = QMailMessageKey::sender(search, QMailDataComparator::Includes);
         QMailMessageKey previewKey = QMailMessageKey::preview(search, QMailDataComparator::Includes);
-        setKey(m_key & (subjectKey | recipientsKey | fromKey | previewKey));
+        m_searchKey = QMailMessageKey(m_key & (subjectKey | recipientsKey | fromKey | previewKey));
+        setKey(m_searchKey);
+
+        m_search = search;
+        // We have model filtering already via searchKey, so we pass just the current model key plus body search,
+        // otherwise results will be merged and just entries with both, fields and body matches will be returned.
+        EmailAgent::instance()->searchMessages(m_key, search, QMailSearchAction::Local, 100);
```
Comment (VDVsx) [inline src/emailmessagelistmodel.cpp]: We add support to define the amount of results to return in next commits.

File: src/emailmessagelistmodel.cpp [hunk 2]
```diff
@@ -924,3 +932,51 @@ void EmailMessageListModel::messagesRemoved(const QMailMessageIdList &ids)
...
+    if (search != m_search) {
+        qCDebug(lcGeneral) << "Search terms are different, skipping. Received: " << search << " Have: " << m_search;
+        return;
+    }
+    switch (status) {
+    case EmailAgent::SearchDone:
+        if (isRemote) {
+            // Append online search results to local ones
+            setKey(key() | QMailMessageKey::id(matchedIds));
+            qCDebug(lcGeneral) << "We have more messages on remote " << remainingMessagesOnRemote;
```
Comment (VDVsx) [inline src/emailmessagelistmodel.cpp]: remainingMessagesOnRemote wil be added as a property here, then users of the model can show a get more/search more on server.

**Comment by VDVsx** [comment]
Update with more commits.

File: src/emailagent.cpp [hunk 3]
```diff
@@ -159,10 +166,13 @@ void EmailAgent::setBackgroundProcess(const bool isBackgroundProcess)
 void EmailAgent::cancelAction(quint64 actionId)
 {
     //cancel running action
-    if (m_currentAction->id() == actionId) {
+    if (m_currentAction && (m_currentAction->id() == actionId)) {
         if(m_currentAction->serviceAction()->isRunning()) {
             m_cancellingSingleAction = true;
             m_currentAction->serviceAction()->cancelOperation();
+        } else {
+            m_currentAction = QSharedPointer<EmailAction>();
```
Comment (matthewvogt) [inline src/emailagent.cpp]: `m_currentAction.reset();`?

File: src/emailagent.cpp [hunk 4]
```diff
@@ -228,6 +238,45 @@ bool EmailAgent::ipcConnected()
...
+
+void EmailAgent::cancelSearch()
+{
+    //cancel running action if is search
+    if (m_currentAction && (m_currentAction->type() == EmailAction::Search)) {
+        if (m_currentAction->serviceAction()->isRunning()) {
+            m_cancellingSingleAction = true;
+            m_currentAction->serviceAction()->cancelOperation();
+        } else {
+            m_currentAction = QSharedPointer<EmailAction>();
```
Comment (matthewvogt) [inline src/emailagent.cpp]: Again.

File: src/emailagent.cpp [hunk 5]
```diff
@@ -228,6 +238,45 @@ bool EmailAgent::ipcConnected()
...
+            processNextAction();
+        }
+    }
+    // Starts from 1 since top of the queue will be removed if above conditions are met.
+    for (int i = 1; i < m_actionQueue.size(); ++i) {
+        if (m_actionQueue.at(i).data()->type() == EmailAction::Search) {
+            m_actionQueue.removeAt(i);
```
Comment (matthewvogt) [inline src/emailagent.cpp]: This will skip the item at (i+1) since i will be incremented in the loop mutator.

Comment (VDVsx) [inline src/emailagent.cpp]: oh, good catch, c&p from another place that is probably also wrong, will change to i++

File: src/emailmessagelistmodel.cpp [hunk 6]
```diff
@@ -325,9 +332,10 @@ int EmailMessageListModel::count() const
 
 void EmailMessageListModel::setSearch(const QString search)
```
Comment (matthewvogt) [inline src/emailmessagelistmodel.cpp]: Should be `const QString &`

File: src/emailmessagelistmodel.cpp [hunk 7]
```diff
@@ -335,9 +343,19 @@ void EmailMessageListModel::setSearch(const QString search)
...
         QMailMessageKey fromKey = QMailMessageKey::sender(search, QMailDataComparator::Includes);
         QMailMessageKey previewKey = QMailMessageKey::preview(search, QMailDataComparator::Includes);
-        setKey(m_key & (subjectKey | recipientsKey | fromKey | previewKey));
+        m_searchKey = QMailMessageKey(m_key & (subjectKey | recipientsKey | fromKey | previewKey));
+        setKey(m_searchKey);
+
+        m_search = search;
+        // We have model filtering already via searchKey, so we pass just the current model key plus body search,
+        // otherwise results will be merged and just entries with both, fields and body matches will be returned.
+        EmailAgent::instance()->searchMessages(m_key, search, QMailSearchAction::Local, m_searchLimit);
```
Comment (matthewvogt) [inline src/emailmessagelistmodel.cpp]: m_searchRemainingOnRemote should get cleared here, I think (and also on cancel?)

**Comment by VDVsx** [comment]
Current review comments addressed.

File: src/emailagent.cpp [hunk 8]
```diff
@@ -228,6 +238,45 @@ bool EmailAgent::ipcConnected()
...
+        if (m_currentAction->serviceAction()->isRunning()) {
+            m_cancellingSingleAction = true;
+            m_currentAction->serviceAction()->cancelOperation();
+        } else {
+            m_currentAction.reset();
+            processNextAction();
+        }
+    }
+    // Starts from 1 since top of the queue will be removed if above conditions are met.
+    for (int i = 1; i < m_actionQueue.size(); i++) {
```
Comment (matthewvogt) [inline src/emailagent.cpp]: I don't think this is right - it needs to be somethign like:
    for (int i = 1; i < m_actionQueue.size(); ) {
        if (...) {
            m_actionQueue.removeAt(i);
        } else {
            ++i;
        }
    }

File: src/emailagent.cpp [hunk 9]
```diff
@@ -1176,7 +1233,7 @@ void EmailAgent::reportError(const QMailAccountId &accountId, const QMailService
 
 void EmailAgent::removeAction(quint64 actionId)
 {
-    for (int i = 0; i < m_actionQueue.size(); ++i) {
+    for (int i = 0; i < m_actionQueue.size(); i++) {
```
Comment (matthewvogt) [inline src/emailagent.cpp]: Same problem as above.

File: src/emailagent.cpp (hunk 8 above)
Comment (VDVsx) [inline src/emailagent.cpp]: Right, we are removing, completely overlooked that, above or QMutableListIterator, but later should be slower.

**Comment by VDVsx** [comment]
Update

**Comment by matthewvogt** [comment]
LGTM

**Pull request merged**